File: 2023/ocaml/day04a/day04a.ml
open Printf

module IntSet = Set.Make(struct
  type t = int
  let compare = compare
end)

let usage () =
  let progname = Sys.argv.(0) in
  printf "usage: %s <file>\n" progname;
  exit 1

let process contents =
  List.fold_left (fun result line ->
      if line <> ""
        then
          (let rest = (List.nth (String.split_on_char ':' line) 1) in
           let winning_str = (List.nth (String.split_on_char '|' rest) 0) in
           let winning_set = List.fold_left (fun set num_str ->
             if num_str <> ""
               then (IntSet.add (int_of_string num_str) set)
               else set
             ) IntSet.empty (String.split_on_char ' ' winning_str) in
           let hand_str = (List.nth (String.split_on_char '|' rest) 1) in
           let hand_set = List.fold_left (fun set num_str ->
               if num_str <> ""
                 then (IntSet.add (int_of_string num_str) set)
                 else set
             ) IntSet.empty (String.split_on_char ' ' hand_str) in
           let intersection = IntSet.inter winning_set hand_set in
           let count = IntSet.cardinal intersection in
           result + if (count > 0) then (Int.shift_left 1 (count - 1)) else 0)
        else result
    ) 0 (String.split_on_char '\n' contents)

let () =
  if Array.length Sys.argv < 2 then
    usage ()
  else
    let filename = Sys.argv.(1) in
    let input_channel = open_in filename in
    try
      let contents = In_channel.input_all input_channel in
      let result = process contents in
        printf "result = %d\n" result
    with e ->
      raise e

File: 2023/go/day04a/day04a.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func usage(progname string) {
	log.Fatalf("usage: %s <file>\n", progname)
	os.Exit(1)
}

func process(contents string) uint32 {
	result := uint32(0)
	lines := strings.Split(contents, "\n")
	for _, line := range lines {
		if len(line) == 0 {
			break
		}
		parts := strings.Split(line, ": ")
		_, rest := parts[0], parts[1]
		restSplit := strings.Split(rest, " | ")
		winningStr, handStr := restSplit[0], restSplit[1]
		winningSet := make(map[uint32]bool)
		winningNums := strings.Split(winningStr, " ")
		for _, numStr := range winningNums {
			if len(numStr) != 0 {
				num, err := strconv.ParseUint(numStr, 10, 32)
				if err != nil {
					panic(err)
				}
				winningSet[uint32(num)] = true
			}
		}
		handNums := strings.Split(handStr, " ")
		handSet := make(map[uint32]bool)
		for _, numStr := range handNums {
			if len(numStr) != 0 {
				num, err := strconv.ParseUint(numStr, 10, 32)
				if err != nil {
					panic(err)
				}
				handSet[uint32(num)] = true
			}
		}
		intersection := make(map[uint32]bool)
		for num := range winningSet {
			if handSet[num] {
				intersection[num] = true
			}
		}
		if len(intersection) > 0 {
			result = result + (1 << (len(intersection) - 1))
		}
	}
	return result
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}
	filename := os.Args[1]
	buf, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}
	contents := string(buf[:])
	result := process(contents)
	fmt.Printf("result = %d\n", result)
}
